ext/gmp/gmp.c

```diff
@@ -106,138 +102,93 @@ typedef struct _gmp_temp {
-		RETURN_THROWS();                                          \
-	}                                                             \
-	temp.is_used = 1;                                             \
-	gmpnumber = temp.num;                                         \
-}
-
-#define FETCH_GMP_ZVAL(gmpnumber, zval, temp, arg_pos)            \
-if (IS_GMP(zval)) {                                               \
-	gmpnumber = GET_GMP_FROM_ZVAL(zval);                          \
-	temp.is_used = 0;                                             \
-} else {                                                          \
-	mpz_init(temp.num);                                           \
-	if (convert_to_gmp(temp.num, zval, 0, arg_pos) == FAILURE) {  \
-		mpz_clear(temp.num);                                      \
-		RETURN_THROWS();                                          \
-	}                                                             \
-	temp.is_used = 1;                                             \
-	gmpnumber = temp.num;                                         \
-}
-
-#define INIT_GMP_RETVAL(gmpnumber) \
-	gmp_create(return_value, &gmpnumber)
-
 static void gmp_strval(zval *result, mpz_t gmpnum, int base);
 static zend_result convert_zstr_to_gmp(mpz_t gmp_number, const zend_string *val, zend_long base, uint32_t arg_pos);
-static zend_result convert_to_gmp(mpz_t gmpnumber, zval *val, zend_long base, uint32_t arg_pos);
-static void gmp_cmp(zval *return_value, zval *a_arg, zval *b_arg, bool is_operator);
 
-/*
- * The gmp_*_op functions provide an implementation for several common types
- * of GMP functions. The gmp_zval_(unary|binary)_*_op functions have to be manually
- * passed zvals to work on, whereas the gmp_(unary|binary)_*_op macros already
- * include parameter parsing.
- */
-typedef void (*gmp_unary_op_t)(mpz_ptr, mpz_srcptr);
-typedef mp_bitcnt_t (*gmp_unary_opl_t)(mpz_srcptr);
+static bool gmp_zend_parse_arg_into_mpz_ex(
+	zval *arg,
+	mpz_ptr *destination_mpz_ptr,
+	uint32_t arg_num,
+	bool is_operator
+) {
+	if (EXPECTED(Z_TYPE_P(arg) == IS_OBJECT)) {
+		if (EXPECTED(instanceof_function(Z_OBJCE_P(arg), gmp_ce))) {
+			*destination_mpz_ptr = GET_GMP_FROM_ZVAL(arg);
+			return true;
+		}
+		return false;
+	}
 
-typedef void (*gmp_unary_ui_op_t)(mpz_ptr, gmp_ulong);
+	*destination_mpz_ptr = GMPG(zpp_arg[arg_num-1]);
+	if (Z_TYPE_P(arg) == IS_STRING) {
+		return convert_zstr_to_gmp(*destination_mpz_ptr, Z_STR_P(arg), /* base */ 0, is_operator ? 0 : arg_num) != FAILURE;
+	}
 
-typedef void (*gmp_binary_op_t)(mpz_ptr, mpz_srcptr, mpz_srcptr);
+	if (Z_TYPE_P(arg) == IS_LONG) {
+		mpz_set_si(*destination_mpz_ptr, Z_LVAL_P(arg));
+		return true;
+	}
 
-typedef void (*gmp_binary_ui_op_t)(mpz_ptr, mpz_srcptr, gmp_ulong);
-typedef void (*gmp_binary_op2_t)(mpz_ptr, mpz_ptr, mpz_srcptr, mpz_srcptr);
-typedef gmp_ulong (*gmp_binary_ui_op2_t)(mpz_ptr, mpz_ptr, mpz_srcptr, gmp_ulong);
+	if (!ZEND_ARG_USES_STRICT_TYPES()) {
```

Comment:
So strict types only apply on function call arguments right and operators operands should always coerce IIRC. But this code doesn't do that or what am I missing? I think there was a discussion about this at one point.